index.ts

```diff
@@ -70,61 +72,91 @@ function appendLog(payload: any) {
     if (!parsedUserData.success) {
-      throw parsedUserData.error;
+      return { error: parsedUserData.error, type: "validation" } as const;
     }
-    await createUser(parsedUserData.data);
 
-    migrated++;
+    return { data: await createUser(parsedUserData.data) } as const;
   } catch (error) {
     if (error.status === 422) {
-      appendLog({ userId: userData.userId, ...error });
-      alreadyExists++;
-      return;
+      return { error, type: "already_exists" } as const;
     }
 
-    // Keep cooldown in case rate limit is reached as a fallback if the thread blocking fails
     if (error.status === 429) {
-      console.log(`Waiting for rate limit to reset`);
-      await cooldown();
-
-      console.log("Retrying");
-      // conditional recursion
-      return processUserToClerk(userData);
+      return { error, type: "rate_limit" } as const;
     }
 
-    appendLog({ userId: userData.userId, ...error });
+    return { error, type: "unknown" } as const;
   }
 }
 
 async function cooldown() {
   await new Promise((r) => setTimeout(r, DELAY));
 }
 
-let spinner: Ora;
-
 async function main() {
-  const parsedUserData = JSON.parse(fs.readFileSync("users.json", "utf-8"));
-
-  console.log(`Clerk User Migration Utility`);
-  spinner = ora(`Migrating users...`).start();
+  const parsedUserData: User[] = JSON.parse(
+    fs.readFileSync("users.json", "utf-8")
+  );
 
-  for (const userData of parsedUserData) {
-    await cooldown();
-    await processUserToClerk(userData);
+  const queue = [...parsedUserData];
+
+  while (queue.length > 0) {
+    const batchedQueue = queue.splice(0, BATCH);
+    let rateLimited = false;
+
+    const migrations = new Listr(
+      batchedQueue.map((user) => ({
+        title: `Migrating user ${user.userId}`,
+        task: async (_, task) => {
+          const result = await processUserToClerk(user);
+
+          if (result.error) {
+            if (result.type === "rate_limit") {
+              task.skip(`Rate limit exceeded, will be retried later`);
+              queue.push(user);
+              rateLimited = true;
+              return;
+            }
+
+            appendLog({ userId: user.userId, ...result.error });
+            task.skip(`Failed to migrate user ${user.userId}`);
+            return;
+          }
+
+          if (result.data) {
+            task.title = `Migrated user ${user.userId}`;
+          }
+
+          return;
+        },
+      })),
+      { concurrent: true }
+    );
+
+    await migrations.run();
+
+    if (rateLimited) {
+      await new Listr([
+        {
+          title: `Rate Limited - cooldown for ${
+            RATE_LIMIT_DELAY / 1000
+          } seconds...`,
+          task: async () => new Promise((r) => setTimeout(r, RATE_LIMIT_DELAY)),
+        },
+      ]).run();
+    } else if (queue.length) {
+      await new Listr([
+        {
+          title: `Cooldown for ${DELAY / 1000} seconds...`,
+          task: cooldown,
+        },
+      ]).run();
+    }
```

Comment:
Once the migrations have finished, cooldown for long delay if we got rate limited, short delay otherwise.
Restart the loop by pulling a new batch from the queue.